File: app/src/main/java/io/github/hrilab/pepper_realtime/TurnManager.java
package io.github.hrilab.pepper_realtime;

import android.util.Log;

public class TurnManager {
    public enum State { IDLE, LISTENING, THINKING, SPEAKING }

    public interface Callbacks {
        void onEnterListening();
        void onEnterThinking();
        void onEnterSpeaking();
        void onExitSpeaking();
    }

    private static final String TAG = "TurnManager";
    private volatile State state = State.IDLE;
    private final Callbacks callbacks;

    public TurnManager(Callbacks callbacks) { this.callbacks = callbacks; }

    public State getState() { return state; }

    public synchronized void setState(State next) {
        if (next == state) return;
        State prev = state;
        state = next;
        Log.d(TAG, "State: " + prev + " -> " + next);
        switch (next) {
            case LISTENING:
                callbacks.onEnterListening();
                break;
            case THINKING:
                callbacks.onEnterThinking();
                break;
            case SPEAKING:
                callbacks.onEnterSpeaking();
                break;
            case IDLE:
                // no-op
                break;
        }
        if (prev == State.SPEAKING && next != State.SPEAKING) {
            callbacks.onExitSpeaking();
        }
    }
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/manager/TurnManager.kt
package io.github.anonymous.pepper_realtime.manager

import android.util.Log

class TurnManager(private var callbacks: Callbacks?) {

    enum class State {
        IDLE, LISTENING, THINKING, SPEAKING
    }

    interface Callbacks {
        fun onEnterListening()
        fun onEnterThinking()
        fun onEnterSpeaking()
        fun onExitSpeaking()
    }

    @Volatile
    var state: State = State.IDLE
        private set

    fun setListener(callbacks: Callbacks?) {
        this.callbacks = callbacks
    }

    @Synchronized
    fun setState(next: State) {
        if (next == state) return
        val prev = state
        state = next
        Log.d(TAG, "State: $prev -> $next")

        when (next) {
            State.LISTENING -> callbacks?.onEnterListening()
            State.THINKING -> callbacks?.onEnterThinking()
            State.SPEAKING -> callbacks?.onEnterSpeaking()
            State.IDLE -> { /* no-op */ }
        }

        if (prev == State.SPEAKING && next != State.SPEAKING) {
            callbacks?.onExitSpeaking()
        }
    }

    companion object {
        private const val TAG = "TurnManager"
    }
}

